src/Engines/BaseEngine.cs

```diff
@@ -451,6 +452,41 @@ public virtual void OnKernelInfoRequest(Message message)
             }
         }
 
+        /// <summary>
+        ///       Called by shell servers to request cancellation of any
+        ///       current execution. This method simply replies with an
+        ///       empty message indicating that the interrupt has been
+        ///       handled. Classes that inherit from <see cref="BaseEngine"/>
+        ///       should override this method, implement custom interrupt
+        ///       handling, and then call this base class method when complete.
+        /// </summary>
+        /// <param name="message">The original request from the client.</param>
+        public virtual void OnInterruptRequest(Message message)
```

Comment:
I think this makes sense. My only thought here would be if we want to have `BaseEngine.Execute` do anything with interrupt requests by default. E.g.: perhaps use this event to trigger a cancellation source that gets passed to the `action` parameter of `BaseEngine.ExecuteAndNotify`?

Reply:
I definitely thought about that, but I'm leaning against it just because of the fact that it would force so much code to change to take the additional parameter, particularly the `Execute()` method of `MagicSymbol` and all derived classes. I suspect many commands won't need or want to support cancellation, such that this could force a lot of changes in consuming code for little benefit.

Reply:
Makes sense!

Reply:
After thinking about this more, I think you're right that having the cancellation source here is the proper way to do it. Yes, it will require changes in consuming code, but I don't see a very clean/correct way for consuming code to handle this on its own that would ensure consistency. I'll make those changes and send a new PR.